### ISSUE TITLE
autogen.gml is blank

### ISSUE THREAD

**Issue by Hyomoto** [YAL-GameMaker/Apollo#1]
I didn't really know where to leave this.  I tried giving compilation a try as you suggested, but no matter what I do, autogen.gml is always blank and it seems like not all the necessary functions end up being compiled for the extension either, such as apollo_init_3.  The BUILD.md file says if I've configured it all correctly it'll do it's magic, but other than not getting any errors, having the files in the right spots, and otherwise *seeming* to be good to go, I'm not sure why it won't generate these files.  I tried just grabbing the file *from* the master and using that, but I end up getting an assertion error.  Could be because indev, probably because not the right extension files.

It should be mentioned that autogen is *created*, it's just blank.  I also tried the release versions of GmxGen and GmlCppExtFuncs as well as compiling them from master but no avail.  In short, I ran out of ideas.  My absolute best guess is that since these files are supposed to be compiled from hints in the other files, that there's some information missing in the files generated by CppExtFuncs but I really have no idea.

Anyways, this isn't an Apollo-specific bug, but if I'm just missing something obvious, please let me know.

**Comment by YellowAfterlife** [comment]
I forgot to add this to BUILD.md, but you need to copy the [Lua source files](https://www.lua.org/download.html) to "Lua" folder since they are not included in the repository. Otherwise the Visual Studio project will not build as such, although I'm not sure how does it end up creating an empty file https://github.com/YAL-GameMaker/Apollo/blob/main/Lua/Put%20your%20Lua%20source%20files%20here.txt

**Comment by Hyomoto** [comment]
I appreciate the reply.  No issue there, unless the version matters. I used 5.4.4 and you can't even build the solution with out them.  
 Went ahead and downgraded Lua just to check, and unsurprisingly that didn't make a difference.  For what it's worth, this part didn't trip me up.  The solution won't build without the files, and you made it pretty clear that they are needed and where they go.  If only all solutions were so easy to work with!  Unfortunately the issue is it compiles, at least on it's face, correctly.  The problem is that the generated extension is ... incomplete?  It's not *blank*, but it's missing most of the functions.  I only know this because the master directory *has* an extension, and that Apollo.yy includes the absent functions.

As mentioned, I also gave that a try, copy the existing project extension and use the new dll, just to see what happens.  This *will* start and run, but throws an assertion error.  I don't think this is much worth dwelling on because it's likely that *those* declarations weren't meant for *this* compilation, but perhaps there is a more obvious clue in that.

**Comment by Hyomoto** [comment]
For completeness, output log.  Maybe something jumps out.  This is *just* the x64 release, I also, as per the instructions, performed the x86 Release build first.
```
1>------ Rebuild All started: Project: Apollo, Configuration: Release x64 ------
1>Running pre-build for Release
1>Running GmlCppExtFuncs...
1>lapi.c
1>lauxlib.c
1>lbaselib.c
1>lcode.c
1>lcorolib.c
1>lctype.c
1>ldblib.c
1>ldebug.c
1>ldo.c
1>ldump.c
1>lfunc.c
1>lgc.c
1>linit.c
1>liolib.c
1>llex.c
1>lmathlib.c
1>lmem.c
1>loadlib.c
1>lobject.c
1>lopcodes.c
1>Compiling...
1>loslib.c
1>lparser.c
1>lstate.c
1>lstring.c
1>lstrlib.c
1>ltable.c
1>ltablib.c
1>ltm.c
1>lundump.c
1>lutf8lib.c
1>lvm.c
1>lzio.c
1>apollo_array_ref.cpp
1>apollo_convert.cpp
1>apollo_global.cpp
1>apollo_init.cpp
1>apollo_interop.cpp
1>apollo_lua_ref.cpp
1>apollo_raw.cpp
1>apollo_shared.cpp
1>apollo_shared.cpp(28,5): warning C4996: 'sprintf': This function or variable may be unsafe. Consider using sprintf_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
1>apollo_struct_ref.cpp
1>autogen.cpp
1>Apollo.cpp
1>gml_api.cpp
1>stdafx.cpp
1>dllmain.cpp
1>   Creating library C:\Users\drago\Desktop\Projects\Apollo-main\x64\Release\Apollo.lib and object C:\Users\drago\Desktop\Projects\Apollo-main\x64\Release\Apollo.exp
1>Generating code
1>Previous IPDB not found, fall back to full compilation.
1>All 1389 functions were compiled because no usable IPDB/IOBJ from previous compilation was found.
1>Finished generating code
1>Apollo.vcxproj -> C:\Users\drago\Desktop\Projects\Apollo-main\x64\Release\Apollo.dll
1>Running post-build for Release
1>Copying documentation...
1>The system cannot find the path specified.
1>        0 file(s) copied.
1>Combining the source files...
1>Running GmxGen...
1>Running GmxGen for "C:\Users\drago\Desktop\Projects\Apollo-main\Apollo_23\extensions\Apollo\Apollo.yy"...
1>Copied "C:\Users\drago\Desktop\Projects\Apollo-main\x64\Release\Apollo.dll"
1>    to "C:\Users\drago\Desktop\Projects\Apollo-main\Apollo_23\extensions\Apollo/Apollo_x64.dll"
1>Checking `autogen.gml`...
1>Checking `apollo_core.gml`...
1>Checking `Apollo.dll`...
1>Done building project "Apollo.vcxproj".
========== Rebuild All: 1 succeeded, 0 failed, 0 skipped ==========
```

**Comment by Hyomoto** [comment]
![image](https://user-images.githubusercontent.com/6281477/218266358-79b7491d-409f-4706-8187-0fe2c8083e6a.png)
And this is the resulting folder.

**Comment by Hyomoto** [comment]
As said above, I don't want you to feel like you need to fix this.  But if it turns out I've made some basic blunder, it's a good learning experience as well as, well, I wouldn't mind playing with the newest version.

**Comment by YellowAfterlife** [comment]
Okay, small correction: `autogen.gml` _should_ be empty as of the current commit - it usually contains GML halves of auto-generated functions, but the extension primarily relies on auto-generating new "mangled" functions (that live entirely on C++ side), so there aren't any of those for now. `autogen.cpp`, on other hand, should have a few hundred lines of weird-looking code in it.

`apollo_init_3` is defined in the YY file - it points to `apollo_init_3_raw` from `autogen.cpp`

But, more importantly, try building for x64 and running the test project - it runs the tests on startup, which is the real way to tell whether the extension is working correctly.

**Comment by Hyomoto** [comment]
Unless I've done something shonky, that's the part I'm at. After building it tells me Apollo_init_3 isn't defined. If I define that it will tell me Lua_state_create_ref_raw doesn't exist, and so on and so forth. If I manually add these they will work, until the next function isn't declared.

That's why I tried the project file in the repository instead which has all the functions, seemingly, declared which *does* start and runs the tests, but fails when asserting a return. Expects 4, gets 0. My guess is because that extension file is out of date, so troubleshooting that seems like wasted effort. Or maybe it does work, and it really is an assertion failure. Still, the build doesn't generate all the expected functions in the extension even though everything outwardly appears to run, so I don't want to tunnel in on "I did something unexpected and got unexpected results" as an indicator.

Just to rule out one more thing, I tried moving the project to another location in case it's some silent OS read permission failure, but no change. This is also why I zoom in on the GMLcpp maybe failing, if the headers that GmxGen are expecting aren't there, it would explain why they aren't getting generated. But, as mentioned, tried the release build and the compiled one with no changes.

One thing that maybe matters is the C# version fails with an index out of bounds error. The build.md gives basic build instructions, but they don't match the GmlCppExtFuncs repository, so I used the project instructions there instead. However, there are two versions and the one generated by Build All.bat crashes. If that's the expected version then it might be I do have the wrong version and would explain why it isn't working. All guesses of course, but it's the only part that doesn't work and I can only reason about what I can see.

**Comment by Hyomoto** [comment]
To follow up on this, I carefully started over and the project files built properly.  I get an assertion error now, but the extension functions all appear present and accounted for.

Anyone hoping for some insights into what I did wrong and then fixed: no idea.  I didn't redownload anything and none of the files changed.  I unzipped the project again, put the lua files in the lua folder, and built the solution.  It's what the instructions says should work and it seems to have.
